### PR TITLE
Re-enable parallel message processing

### PR DESCRIPTION
**Pull request by caohai**
Follow-up PR for https://github.com/microsoft/sqltoolsservice/pull/1596 and https://github.com/microsoft/sqltoolsservice/pull/1655

This PR adds back the parallel message processing and adds an additional flag at handler level to control if this handler supports running in parallel.

Some handlers are already using `Task.Run()` so they are good without this change.

If the user doesn't choose to enable the `ParallelMessageProcessing` flag, the behavior should be exactly the same as the behavior without this change.

| Service Name      | Marked As Supporting Parallel Processing | Already using Task.Run() |Comment|
| ----------- | ----------- |----------- |-|
|AdminService.cs |YES|YES|Enabling(already async)|
|AgentService.cs |YES|NO|Enabling (no concern for running in parallel)|
|BlobService.cs |YES|NO|Enabling (no concern for running in parallel)|
|AzureFunctionsService.cs |YES|NO|Enabling (no concern for running in parallel)|
|CmsService.cs |YES |YES|Enabling(already async)|
|ConnectionService.cs |YES|YES|Enabling(already async)|
|DacFxService.cs |YES|YES|Enabling(already async)|
|DisasterRecoveryService.cs |YES|YES|Enabling(already async)|
|EditDataService.cs |YES|NO|Enabling (no concern for running in parallel)|
|ExecutionPlanService.cs |YES|NO|Enabling (no concern for running in parallel)|
|FileBrowserService.cs |YES|YES|Enabling(already async)|
|TSqlFormatterService.cs |YES|YES|Enabling(already async)|
|ExternalLanguageService.cs |YES|NO|Enabling (no concern for running in parallel)|
|LanguageService.cs |NO||order of msgs can matter, example: HandleTokenRefreshedNotification|
|MetadataService.cs |YES|NO|Enabling (no concern for running in parallel)||
|MigrationService.cs |YES|NO|Enabling (no concern for running in parallel)||
|ModelManagementService.cs |YES|NO|Enabling (no concern for running in parallel)||
|NotebookConvertService.cs |YES|NO|Enabling (no concern for running in parallel)||
|ObjectExplorerService.cs |NO|YES|Not enabling as OE could hang||
|ObjectManagementService.cs |YES|NO|Enabling (no concern for running in parallel)|
|ProfilerService.cs |YES|NO|Enabling (no concern for running in parallel)|
|QueryExecutionService.cs |YES|NO| Enabling(already async)|
|SerializationService.cs |YES|YES|Enabling(already async)|
|SchemaCompareService.cs |YES|YES|Enabling(already async)|
|ScriptingService.cs |YES|YES|Enabling(already async)|
|SecurityService.cs |YES|YES|Enabling(already async)|
|ServerConfigService.cs |YES|NO|Enabling (no concern for running in parallel)|
|SqlAssessmentService.cs |YES|YES|Enabling(already async)|
|TableDesignerService.cs |YES|YES|Enabling(already async)|
|TaskService.cs |YES|YES|Enabling(already async)|
|WorkspaceService.cs |NO|   |Not enabling as there could be doc out of sync error |

### CODE REVIEW

**Comment by alanrenmsft** [comment]
@caohai  Thanks for following up on this. Could you please add another column Reason explaining why it is not enabled/enabled? also, for the services/handlers do not support parallel processing, should also add comment to the code.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/IProtocolEndpoint.cs [hunk 1]
```diff
@@ -18,15 +18,18 @@ public interface IProtocolEndpoint : IEventSender, IRequestSender
     {
         void SetRequestHandler<TParams, TResult>(
             RequestType<TParams, TResult> requestType,
-            Func<TParams, RequestContext<TResult>, Task> requestHandler);
+            Func<TParams, RequestContext<TResult>, Task> requestHandler,
+            bool isParallelProcessingSupported = false);
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/IProtocolEndpoint.cs:22]: Please add doc comments for these methods and at the very least add a param comment for the new param being added (but if you could do the others while you're here that'd be great)

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 2]
```diff
@@ -27,14 +27,22 @@ public class MessageDispatcher
...
 
+        internal Dictionary<string, bool> eventHandlerParallelismMap =
+            new Dictionary<string, bool>();
+
         private Action<Message> responseHandler;
 
         private CancellationTokenSource messageLoopCancellationToken =
             new CancellationTokenSource();
 
+        private SemaphoreSlim semaphore = new SemaphoreSlim(10); // Limit to 10 threads
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:44]: Please add the reason why 10 was chosen to the comment

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 3]
```diff
@@ -27,14 +27,22 @@ public class MessageDispatcher
         internal Dictionary<string, Func<Message, MessageWriter, Task>> requestHandlers =
             new Dictionary<string, Func<Message, MessageWriter, Task>>();
 
+        internal Dictionary<string, bool> requestHandlerParallelismMap =
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:30]: Instead of having a separate dictionary just to keep track of bools I would suggest using a tuple type in the requestHandlers dictionary

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples

`internal Dictionary<string, (Func<Message, MessageWriter, Task>, bool)> requestHandlers =
            new Dictionary<string, (Func<Message, MessageWriter, Task>, bool)>();`

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 4]
```diff
@@ -324,13 +354,15 @@ private async Task ListenForMessages(CancellationToken cancellationToken)
 
             if (handlerToAwait != null)
             {
-                if (this.ParallelMessageProcessing)
+                if (this.ParallelMessageProcessing && isParallelProcessingSupported)
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:357]: This currently means that any non-parallel enabled messages will block all other messages from being processed until it's completed. But that means we're only getting some of the benefit of going parallel in the first place - ideally we should be having all parallel messages kicked off as soon as they're read in while any non-parallel ones will be handled sequentially.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 5]
```diff
@@ -324,13 +354,15 @@ private async Task ListenForMessages(CancellationToken cancellationToken)
...
                 {
                     // Run the task in a separate thread so that the main
                     // thread is not blocked.
+                    await semaphore.WaitAsync();
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:361]: I'd suggest adding a comment explaining why you're doing this.

**Review by Charles-Gagnon — COMMENTED** [review]
I'm quite wary of bringing this back without knowing what the problem was in the first place. Especially with so many other huge changes going in this month already.

1. Are you planning on having this in the November release?
2. Have you done any more investigation into what the original problem was?
3. What's the plan for if we start seeing more issues?

Given that we've lived with the current functionality this entire time I don't see this being something we need to rush back into until we have more confidence there aren't regressions, so my suggestion is we work some more on building up that confidence before moving ahead with this.

**Comment by alanrenmsft** [comment]
@Charles-Gagnon makes sense, @caohai let's do it after the Nov release so that we could ask the user to try the insider build and see whether they could reproduce the issue with Insider builds.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 6]
```diff
@@ -112,43 +120,55 @@ public void Stop()
...
+                        catch (Exception ex)
+                        {
+                            Logger.Error($"{requestType.MethodName} : {ex}");
+                            await requestContext.SendError(ex.Message);
                         }
+                    };
 
-                        await requestHandler(typedParams, requestContext);
+                    if (!this.ParallelMessageProcessing || !isParallelProcessingSupported) {
+                        await func();
```
Comment (caohai) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:167]: This part is to make sure if a user does not explicitly turn on the `ParallelMessageProcessing` flag, the message handling behavior will remain unchanged.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs (hunk 3 above)
Comment (caohai) [reply src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:30]: I gave this a try but personally the readability feels worse to me (even with the use of field name properties), I might stick with the additional dict here.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 7]
```diff
@@ -166,39 +186,52 @@ public void Stop()
...
-                    try
-                    {                
-                        if (eventMessage.Contents != null)
-                        {
-                            try
-                            {
-                                typedParams = eventMessage.Contents.ToObject<TParams>();
-                            }
-                            catch (Exception ex)
+                    Func<Task> func = async () =>
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:203]: What you're defining here with `async (eventMessage, messageWriter) => { }` is already a function, why are you creating another function `Func<Task>` inside and awaiting on it?

Same comment for above method.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 8]
```diff
@@ -324,28 +360,31 @@ private async Task ListenForMessages(CancellationToken cancellationToken)
...
+                if (this.ParallelMessageProcessing && isParallelProcessingSupported)
                 {
                     // Run the task in a separate thread so that the main
-                    // thread is not blocked.
+                    // thread is not blocked. Use semaphore to limit the degree of parallelism.
+                    await semaphore.WaitAsync();
                     _ = Task.Run(() =>
                     {
-                        _ = RunTask(handlerToAwait);
+                        _ = RunTask(handlerToAwait, messageToDispatch);
```
Comment (cheenamalhotra) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:370]: Shouldn't this be awaited before releasing semaphore?

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs (hunk 4 above)
Comment (caohai) [reply src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:363]: Yes we are only getting part of the benefit here. I attempted to get msg loop on its own thread but that will be a big change so we are going this route for now.

**Comment by alanrenmsft** [comment]
@caohai please work with @cheenamalhotra and make sure the logs can capture the query hang issue she is facing.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs (hunk 7 above)
Comment (cheenamalhotra) [reply src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:203]: Pasting a simple example for reference, no code would be executed until you await on the function, as this is an async task object you're passing around.

```
Func<int, Task> func = async(i) =>
{
    Console.WriteLine("2. Non awaited result");
    await Console.Out.WriteLineAsync("3. Awaited result");
};
var function = func;
await Task.Delay(100);
await Console.Out.WriteLineAsync("1. Before await");
await function(1);
```

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Messaging/MessageDispatcherTests.cs [hunk 9]
```diff
@@ -84,5 +86,67 @@ public void OnListenTaskCompletedFaultedTaskTest()
...
+            bool okayToEnd = false;
+            mockChannel.Setup(c => c.MessageReader.ReadMessage())
+                .Returns(Task.FromResult(Message.Request("1", "test/requestType", null)));
+            var dispatcher = new MessageDispatcher(mockChannel.Object);
+            dispatcher.ParallelMessageProcessing = parallelMessageProcessing;
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            var handler = async (int _, RequestContext<int> _) =>
+            {
+                Thread.Sleep(msForEachRequest);
+                unfinishedRequestCount.Wait();
```
Comment (cheenamalhotra) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Messaging/MessageDispatcherTests.cs:123]: What's the purpose of waiting sync here?
You can await with `WaitAsync()` on it.

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs (hunk 7 above)
Comment (cheenamalhotra) [reply src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:203]: Also tested the original code from @alanrenmsft , it doesn't have negative consequences from execution perspective. 
It works normally as expected:

```
Func<int, Task> func = (i) => Task.Run(async () => 
{
    Console.WriteLine("2. Non awaited result");
    await Console.Out.WriteLineAsync("3. Awaited result");
});
var function = func;
await Task.Delay(100);
await Console.Out.WriteLineAsync("1. Before await");
await function(1);
```

Comment (caohai) [reply src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:203]: The difference here is that the function actually gets invoked here with `handlerToAwait = requestHandler(messageToDispatch, messageWriter);`, I'll move this call to where await happens.

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Messaging/MessageDispatcherTests.cs [hunk 10]
```diff
@@ -84,5 +86,67 @@ public void OnListenTaskCompletedFaultedTaskTest()
...
+                    stopwatch.Stop();
+                    okayToEnd = true;
+                }
+                await Task.CompletedTask;
+            };
+
+            dispatcher.SetRequestHandler(requestType, handler, false, true);
+            dispatcher.Start();
+
+            while (true)
```
[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply test/Microsoft.SqlTools.ServiceLayer.UnitTests/Messaging/MessageDispatcherTests.cs:137]: This can be ignored, thanks for meeting offline and explaining the requirement!

File: src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs [hunk 11]
```diff
@@ -284,19 +300,13 @@ private async Task ListenForMessages(CancellationToken cancellationToken)
             Message messageToDispatch,
             MessageWriter messageWriter)
         {
-            Task handlerToAwait = null;
+            Func<Message, MessageWriter, Task> handlerToAwait = null;
```
Comment (caohai) [inline src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs:303]: changed hanlderToAwait to a Func instead of a Task and removed `handlerToAwait = requestHandler(messageToDispatch, messageWriter);` so we don't have to wait for the potential slow sync code in the handler. The invocation of hanlderToAwait happens later when we awaits.

**Comment by caohai** [comment]
@alanrenmsft This PR should be ready for another review now.

**Review by alanrenmsft — APPROVED** [review]

**Comment by alanrenmsft** [comment]
there are still test failures.

**Comment by caohai** [comment]
> nfinishedRequestCount.Wait();

The test failed was the one I added, the assertion is a bit strict and can become flaky if the build machine has small amount of logical cores. I'll take a look to loose the assertion a bit.
